gptel.el

```diff
@@ -1991,6 +1997,97 @@ buffer."
+
+(defcustom gptel-augment-post-modify-hook nil
+  "Hook run after augmentally extending the context of a gptel request.
+
+This runs (possibly) before any request is submitted."
+  :type 'hook)
+
+(defcustom gptel-augment-handler-functions nil
+  "List of handlers used to augment a query before sending it.
+
+Each handler function receives a `callback' and the current
+`fsm'.  When completed, it must call that `callback' with the
+final version of that `fsm' for the next stage of the
+augmentation pipeline.
+
+Note that the `info' of this `fsm' is special in that it is the
+raw, unprocessed input, before the `:data' has been constructed
+by the backends. The `:data' in this version of the info is a
+pre-construction set of arguments, all of which may be examined
+and/or modified by the augment handler in order to influence
+construction of the final request data. This pre-construction set
+has the following elements:
+
+  (list :args t
+        :full-prompt full-prompt
+        :use-context use-context
+        :system-message system-message
+        :collection collection
+        :stream stream
+        :callback callback
+        :context context
+        :in-place in-place
+        :gptel-include-reasoning gptel-include-reasoning
+        :gptel-use-tools gptel-use-tools)
+
+Note that while this set of handlers can certainly be set with a
+global value to be applied to all queries in all buffers, it
+meant to be set locally for a specific buffer, or chat topic, or
+only the context of using a certain agent."
+  :type 'hook)
+
+(defmacro gptel-sync-fn (fn)
+  "Turn a synchronous function into a suitable augment handler.
+
+This takes any function from FSM -> FSM, and returns a function
+that takes a callback and an FSM, and once it finishes by
+returning the next fsm, calls the callback on that value.
+
+Thus, you can take a normal, synchronous augmentation function
+and add it to `gptel-augment-handler-functions' using the
+following code:
+
+  (add-hook 'gptel-augment-handler-functions
+            (gptel-sync-fn #'my-synchronous-function))"
+  `(lambda (callback fsm)
+     (funcall callback (funcall fn fsm))))
+
+(defun gptel--finish-augmentation (fsm)
+  "The last augmentation to be called; transitions the FSM."
+  (setf (gptel-fsm-info fsm)
+        (gptel--realize-info (gptel-fsm-info fsm)))
+  (gptel--fsm-transition fsm)
+  (run-hooks 'gptel-augment-post-modify-hook))
+
+(defun gptel--augment-info (fsm fns)
+  "Augment the request contained in state machine FSM's info.
+
+Each function receives a CALLBACK and an INFO, and must call the
+callabck with their augemented version of the INFO (or the
+original INFO) once completed. This permits augmentations to be
+asynchronous.
+
+Returns the transformed info at the end."
+  (if (null fns)
+      (gptel--finish-augmentation fsm)
+    (funcall (car fns)
+             #'(lambda (fsm) (gptel--augment-info fsm (cdr fns)))
+             fsm)))
+
+(defun gptel--handle-augment (fsm)
```

Comment:
If we switch to a counting callback in `gptel--augment-info` (see comment), we don't need a (recursive) `gptel--augment-info` or `gptel--finish-augmentation` any more, and  `gptel--handle-augment` can handle the whole augment process, localizing all the augment code.